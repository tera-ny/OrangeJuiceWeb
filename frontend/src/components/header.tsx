import { FC } from "react";
import Link from "next/link";
import {
  isSubscribedState,
  uidState,
  useToggleAuthModalRequest,
} from "~/stores/auth";
import { useRecoilValue } from "recoil";
import { profileState } from "~/stores/profile";

const Header: FC = () => {
  const toggle = useToggleAuthModalRequest();
  const { currentUser } = useRecoilValue(profileState);
  const uid = useRecoilValue(uidState);
  const isSubscribed = useRecoilValue(isSubscribedState);
  return (
    <>
      <header className="container">
        <Link href={"/"}>
          <a className="link">
            <picture>
              <source
                srcSet="/img/logo_full_dark.svg"
                media="(prefers-color-scheme: dark)"
              />
              <img
                className="logo"
                height="60"
                width="220"
                src="/img/logo_full_light.svg"
                alt="raspberry-juice"
              />
            </picture>
          </a>
        </Link>
        {isSubscribed
          ? (uid
            ? <span>{currentUser?.name}</span>
            : (
              <button onClick={() => toggle()}>
                Login
              </button>
            ))
          : null}
      </header>
      <style jsx>
        {`
        .container {
          padding: 20px 24px;
          display: flex;
          justify-content: space-between;
        }
        .logo {
          height: 40px;
          width: auto;
        }
        .link {
          display: inline-block;
          height: 40px;
        }
        @media screen and (max-width: 899px) {
          .container {
            padding: 8px 12px;
          }
        }
      `}
      </style>
    </>
  );
};

export default Header;
