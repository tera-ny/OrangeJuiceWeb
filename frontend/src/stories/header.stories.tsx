import { ComponentMeta, ComponentStory } from "@storybook/react";
import { Timestamp } from "firebase/firestore";
import { useEffect } from "react";
import { useSetRecoilState } from "recoil";
import Header from "~/components/header";
import { recoilProvider } from "~/decorators/recoil";
import { authState } from "~/stores/auth";
import { profileState } from "~/stores/profile";

type Story = ComponentStory<typeof Header>;
type Meta = ComponentMeta<typeof Header>;

export default {
  component: Header,
  decorators: [recoilProvider],
} as Meta;

export const Default: Story = () => {
  return <Header />;
};

export const ShouldLogin: Story = () => {
  const setAuth = useSetRecoilState(authState);
  useEffect(() => {
    setAuth({ subscription: {}, openAuthModalRequest: false });
  }, []);
  return <Header />;
};

export const LoggedIn: Story = () => {
  const setAuth = useSetRecoilState(authState);
  const setProfile = useSetRecoilState(profileState);
  useEffect(() => {
    setAuth({ subscription: { uid: "xxxx" }, openAuthModalRequest: false });
    setProfile({
      currentUser: {
        name: "logged in user",
        description: "",
        isActive: true,
        isBanned: false,
        registeredAt: Timestamp.now(),
        theme: 0,
      },
    });
  }, []);
  return <Header />;
};
