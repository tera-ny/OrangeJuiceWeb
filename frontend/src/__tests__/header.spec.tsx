import * as stories from "~/stories/header.stories";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { composeStory } from "@storybook/react";

const cases: [stories: keyof typeof stories][] = [
  ["Default"],
  ["ShouldLogin"],
  ["LoggedInUser"],
];

describe("Header component", () => {
  describe.each(cases)("%s", (story) => {
    const Story = composeStory(stories[story], stories.default);
    it("render component", () => {
      const { container } = render(<Story />);
      expect(container.firstChild).toMatchSnapshot();
    });
    if (story === "LoggedInUser") {
      it("displayed user name", () => {
        const { queryByText } = render(<Story />);
        expect(queryByText("logged in user")).not.toBeNull();
      });
    }
    if (story === "ShouldLogin") {
      it("displayed login button", () => {
        const { queryByRole, queryByText } = render(<Story />);
        expect(queryByText("logged in user")).toBeNull();
        expect(queryByRole("button", { name: "Login" })).not.toBeNull();
      });
    }
  });
});
