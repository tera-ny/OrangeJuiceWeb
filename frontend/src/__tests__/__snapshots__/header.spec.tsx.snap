// Vitest Snapshot v1

exports[`Header component > Default > render component 1`] = `
<header
  class="container"
>
  <a
    class="link"
    href="/"
  >
    <picture>
      <source
        media="(prefers-color-scheme: dark)"
        srcset="/img/logo_full_dark.svg"
      />
      <img
        alt="raspberry-juice"
        class="logo"
        height="60"
        src="/img/logo_full_light.svg"
        width="220"
      />
    </picture>
  </a>
</header>
`;

exports[`Header component > LoggedInUser > render component 1`] = `
<header
  class="container"
>
  <a
    class="link"
    href="/"
  >
    <picture>
      <source
        media="(prefers-color-scheme: dark)"
        srcset="/img/logo_full_dark.svg"
      />
      <img
        alt="raspberry-juice"
        class="logo"
        height="60"
        src="/img/logo_full_light.svg"
        width="220"
      />
    </picture>
  </a>
  <span>
    logged in user
  </span>
</header>
`;

exports[`Header component > ShouldLogin > render component 1`] = `
<header
  class="container"
>
  <a
    class="link"
    href="/"
  >
    <picture>
      <source
        media="(prefers-color-scheme: dark)"
        srcset="/img/logo_full_dark.svg"
      />
      <img
        alt="raspberry-juice"
        class="logo"
        height="60"
        src="/img/logo_full_light.svg"
        width="220"
      />
    </picture>
  </a>
  <button>
    Login
  </button>
</header>
`;
